### DISCUSSION TITLE
Audio Feature Extraction using Whisper

### DISCUSSION THREAD

**Discussion by pegahsalehi** [openai/whisper#1246] [Q&A] [answered]
Hi,
I want to use the whisper to extract logits from audio using `speechbrain.lobes.models.huggingface_whisper`

```
from speechbrain.lobes.models.huggingface_whisper import HuggingFaceWhisper
import speechbrain as sb

model_hub_whisper = "openai/whisper-tiny"
model_whisper = HuggingFaceWhisper(model_hub_whisper, save_path='----')

source = sb.dataio.dataio.read_audio('./intro.wav').squeeze()
source = source.unsqueeze(0)
print(source.shape)
# This can be given as an argument when we instantiate the model as well
model_whisper.encoder_only=True
fea_whisper = model_whisper(source)
print(fea_whisper.shape)
```

how can I extract the logits from audio in the shape` (X, Y)`. 
That `X` is a fixed number when we use the whisper model and `Y` depends on the length of audio.

**Comment by jongwook** [comment, marked as answer]
I presume you mean audio features rather than logits since you're using only the encoder. The encoder always takes 30-second-long audio as input, and we trim or pad the audio to match this length. The encoded features are also 30-seconds long as a result. You can slice the features if the input was shorter than 30 seconds.

More code-specific questions could be better answered by [speechbrain's maintainers](https://github.com/speechbrain/speechbrain).

**Reply by pegahsalehi** [reply]
> I presume you mean audio features rather than logits since you're using only the encoder. The encoder always takes 30-second-long audio as input, and we trim or pad the audio to match this length. The encoded features are also 30-seconds long as a result. You can slice the features if the input was shorter than 30 seconds.
> 
> More code-specific questions could be better answered by [speechbrain's maintainers](https://github.com/speechbrain/speechbrain).

Hi @jongwook , Thanks a lot for your help.
How can I extract logits in the shape `(X, Y)`?

**Reply by LiuRicky** [reply]
> I presume you mean audio features rather than logits since you're using only the encoder. The encoder always takes 30-second-long audio as input, and we trim or pad the audio to match this length. The encoded features are also 30-seconds long as a result. You can slice the features if the input was shorter than 30 seconds.
> 
> More code-specific questions could be better answered by [speechbrain's maintainers](https://github.com/speechbrain/speechbrain).

 @jongwook thanks for your reply, you mentioned that 'You can slice the features if the input was shorter than 30 seconds.'

I have a 6-second audio and input it to whisper encoder. The 6-second audio is pad to 30s, and the final feature shape is (1500, 1024). Can I just slice (:300, :) as my audio feature? 

However, I print the features at (610:620, :), it is non-zero and I think these padded feature indeed involved in encoder transformer since 'mask is ignored in whisper encoder'. Based on these findings, can I just slice (:300, :) as my audio feature? 
![f5fae58f21109fb21112e2bd7567e6f](https://github.com/user-attachments/assets/f4166c24-3dcc-478a-98fc-0bf4a5152905)
![image](https://github.com/user-attachments/assets/925b94cf-2f26-445e-90a1-21b701b8883f)

**Reply by LiuRicky** [reply]
Will the slice incur feature loss since padded feature are involved in the whisper encoder?

**Comment by pegahsalehi** [comment]
@LiuRicky, this is exactly my problem. Did you find any ideas?

**Comment by jeeyung** [comment]
I am also facing the same problem
